aten/src/ATen/native/mps/operations/BinaryOps.mm

```diff
@@ -51,7 +55,17 @@ void binaryOpTensor(const Tensor& self_t, const Tensor& other_t, const Tensor& o
           newCachedGraph = new BinaryOpCachedGraph(mpsGraph);
           newCachedGraph->primaryTensor   = mpsGraphRankedPlaceHolder(mpsGraph, self_dtype , getMPSShape(self));
           newCachedGraph->secondaryTensor = mpsGraphRankedPlaceHolder(mpsGraph, other_dtype, getMPSShape(other));
-          newCachedGraph->outputTensor = binaryBlock(mpsGraph, newCachedGraph->primaryTensor, newCachedGraph->secondaryTensor);
+          MPSGraphTensor* primary = nil;
+          MPSGraphTensor* secondary = nil;
+          if(is_arithmetic && self_dtype != output_dtype)
+            primary = [mpsGraph castTensor:newCachedGraph->primaryTensor toType:output_dtype name:@"primary"];
```

Comment:
Why do you even have to cast when you could create the placeholder with the desired type in mpsGraphRankedPlaceholder()? I mean, In your approach you create a ranked place holder with some type, and the immediately cast it to another. So why not put that logic to compute the correct data type first (like I did in the code you commented in 'self_dtype'), and then pass it to mpsGraphRankedPlaceHolder(), instead of doing it inefficiently on two phases.

Reply:
If the input tensors are say Bool and Float, then I assumed that creating a Float placeholder for the Bool tensor wouldn't work, I'll try it though

Reply:
If I set the placeholder type to (for example) float, when the input tensor is bool, I get this error:

failed assertion `Incompatible element type for parameter at index 1, mlir module expected element type f32 but received i1'